crates/core/tedge_agent/src/software_manager/actor.rs

```diff
@@ -52,48 +74,68 @@ impl Actor for SoftwareManagerActor {
-        .map_err(|err| RuntimeError::ActorError(Box::new(err)))?;
+        let mut message_box = self
+            .message_box
+            .take()
+            .expect("Can't run this actor more than once");
 
-        if plugins.empty() {
-            warn!(
-                "{}",
-                NoPlugins {
-                    plugins_path: self.config.sm_plugins_dir.clone(),
-                }
-            );
-        }
+        self.process_pending_sm_operation(&mut message_box).await?;
+
+        let (output_sender, receiver) = message_box.split();
+        let (mut input_receiver, mut signal_receiver) = receiver.split();
+
+        let software_manager = SoftwareManager {
+            config: Arc::clone(&self.config),
+            state_repository: Arc::clone(&self.state_repository),
+            response_sender: output_sender.sender_clone(),
+        };
 
-        self.process_pending_sm_operation().await?;
+        let mut processing_loop = tokio::spawn(async move {
+            while let Some(request) = input_receiver.next().await {
+                let operation_logs = operation_logs.clone();
 
-        while let Some(request) = self.message_box.recv().await {
-            match request {
-                SoftwareRequest::SoftwareUpdateRequest(request) => {
-                    if let Err(err) = self
-                        .handle_software_update_operation(&request, &mut plugins, &operation_logs)
-                        .await
-                    {
-                        error!("{:?}", err);
+                let result = match request {
+                    SoftwareRequest::SoftwareUpdateRequest(request) => {
+                        software_manager
+                            .handle_software_update_operation(&request, &operation_logs)
+                            .await
                     }
-                }
-                SoftwareRequest::SoftwareListRequest(request) => {
-                    if let Err(err) = self
-                        .handle_software_list_operation(&request, &plugins, &operation_logs)
-                        .await
-                    {
-                        error!("{:?}", err);
+                    SoftwareRequest::SoftwareListRequest(request) => {
+                        software_manager
+                            .handle_software_list_operation(&request, &operation_logs)
+                            .await
                     }
+                };
+
+                if let Err(err) = result {
+                    error!("{err:?}");
                 }
             }
+        });
+
+        tokio::select! {
+            // If the function processing a single SoftwareRequest were
+            // cancel-aware, we could wait for it to return an error response
+            // via the channel. As of now, it wouldn't really change anything,
+            // because most likely the source of the SoftwareRequest has already
+            // terminated as well, so for now we just terminate.
```

Comment:
If there was a mechanism for the task to know if it should cancel, e.g. because the actor needs to be shut down. This could be achieved by, as you have proposed below, using a cancellation token, or even perhaps a plain oneshot channel.

The tricky part may be to decide how deep do we want to pass the cancellation token. E.g. in the case of Software Manager Actor, we might want to pass the cancellation token to a function that handles a single `SoftwareRequest`, so that it can send a proper response when it's cancelled.

However for an operation like download, on cancellation, what should we do with the partially downloaded file? One option is for the downloader to assume that the caller can't do anything with an aborted file, and delete it on abort. But maybe that's not the case and it'd be better for the caller to decide what to do with the file. If so, there is no need for the download to be cancel-aware, because there's no cleanup to run.

Reply:
> If there was a mechanism for the task to know if it should cancel, e.g. because the actor needs to be shut down. This could be achieved by, as you have proposed below, using a cancellation token, or even perhaps a plain oneshot channel.

For a centralized shutdown using cancellation token might be more appropriate as the runtime will not have to maintain a channel per actor.

> The tricky part may be to decide how deep do we want to pass the cancellation token. E.g. in the case of Software Manager Actor, we might want to pass the cancellation token to a function that handles a single `SoftwareRequest`, so that it can send a proper response when it's cancelled.

If all the actors are given each a cancellation token (when built or along the `run()` method), then each actor can uses this token (or a clone) at different internal level. But I would keep things simple here with a whole software operation request cancellation.

> However for an operation like download, on cancellation, what should we do with the partially downloaded file? One option is for the downloader to assume that the caller can't do anything with an aborted file, and delete it on abort. But maybe that's not the case and it'd be better for the caller to decide what to do with the file. If so, there is no need for the download to be cancel-aware, because there's no cleanup to run.

We can live with the second option till we have an improved cancellation story. In any case, both the downloader and the caller must be prepared that some garbage has been left by a previously cancelled request.